File: static/js/alerts.js
// ======================================================
// FILE     : static/js/alerts.js
// PURPOSE  : Fetches alerts from Flask API and updates
//            dashboard badge counts + alert panel
// ======================================================

const FLASK_API   = 'http://localhost:5000';
const REFRESH_MS  = 60000; // refresh every 60 seconds

// ======================================================
// MAIN — fetch and render alerts
// ======================================================

function fetchAlerts() {
    fetch(`${FLASK_API}/api/alerts/`)
        .then(res => res.json())
        .then(data => {
            renderBadges(data);
            renderPanel(data);
        })
        .catch(err => {
            // Flask API not running — fail silently
            console.warn('REDS Alerts: Flask API not reachable.', err);
        });
}


// ======================================================
// BADGES — update count bubbles in topbar / sidebar
// ======================================================

function renderBadges(data) {
    const redCount    = data.red_count    || 0;
    const yellowCount = data.yellow_count || 0;

    // Topbar bell badge
    const bell = document.getElementById('alert-badge-count');
    if (bell) {
        const total = redCount + yellowCount;
        bell.textContent = total;
        bell.style.display = total > 0 ? 'inline-block' : 'none';
        bell.className = redCount > 0
            ? 'badge bg-danger rounded-pill'
            : 'badge bg-warning text-dark rounded-pill';
    }

    // Sidebar red badge
    const redBadge = document.getElementById('sidebar-red-count');
    if (redBadge) {
        redBadge.textContent = redCount;
        redBadge.style.display = redCount > 0 ? 'inline-block' : 'none';
    }

    // Sidebar yellow badge
    const yellowBadge = document.getElementById('sidebar-yellow-count');
    if (yellowBadge) {
        yellowBadge.textContent = yellowCount;
        yellowBadge.style.display = yellowCount > 0 ? 'inline-block' : 'none';
    }

    // Green info counts
    const availablePlots = document.getElementById('info-available-plots');
    if (availablePlots && data.green) {
        availablePlots.textContent = data.green.available_plots || 0;
    }

    const receiptsToday = document.getElementById('info-receipts-today');
    if (receiptsToday && data.green) {
        receiptsToday.textContent = data.green.receipts_today || 0;
    }
}


// ======================================================
// PANEL — render full alert detail in dropdown
// ======================================================

function renderPanel(data) {
    const panel = document.getElementById('alert-panel-body');
    if (!panel) return;

    let html = '';

    // ── RED: Overdue installments ──────────────────────
    if (data.red && data.red.overdue_installments.length > 0) {
        html += `<div class="alert-section">
            <div class="alert-section-title text-danger">
                <i class="bi bi-exclamation-circle-fill"></i>
                Overdue Installments (${data.red.overdue_installments.length})
            </div>`;
        data.red.overdue_installments.forEach(item => {
            html += `<div class="alert-item border-start border-danger border-3 ps-2 mb-2">
                <div class="fw-semibold">${item.customer_name}</div>
                <div class="text-muted small">
                    Plot ${item.plot_no} — ${item.block_name} |
                    Installment #${item.installment_no} |
                    Due: ${item.due_date}
                </div>
                <div class="text-danger small fw-bold">
                    Rs. ${Number(item.amount).toLocaleString()}
                </div>
            </div>`;
        });
        html += `</div>`;
    }

    // ── RED: Bounced cheques ───────────────────────────
    if (data.red && data.red.bounced_cheques.length > 0) {
        html += `<div class="alert-section mt-2">
            <div class="alert-section-title text-danger">
                <i class="bi bi-x-circle-fill"></i>
                Bounced Cheques (${data.red.bounced_cheques.length})
            </div>`;
        data.red.bounced_cheques.forEach(item => {
            html += `<div class="alert-item border-start border-danger border-3 ps-2 mb-2">
                <div class="fw-semibold">${item.customer_name}</div>
                <div class="text-muted small">
                    Cheque #${item.cheque_no} — ${item.cheque_bank} |
                    Receipt: ${item.receipt_no}
                </div>
                <div class="text-danger small fw-bold">
                    Rs. ${Number(item.amount).toLocaleString()}
                </div>
            </div>`;
        });
        html += `</div>`;
    }

    // ── RED: Pending refunds ───────────────────────────
    if (data.red && data.red.pending_refunds.length > 0) {
        html += `<div class="alert-section mt-2">
            <div class="alert-section-title text-danger">
                <i class="bi bi-arrow-counterclockwise"></i>
                Pending Refunds (${data.red.pending_refunds.length})
            </div>`;
        data.red.pending_refunds.forEach(item => {
            html += `<div class="alert-item border-start border-danger border-3 ps-2 mb-2">
                <div class="fw-semibold">${item.customer_name}</div>
                <div class="text-muted small">
                    Plot ${item.plot_no} — ${item.block_name} |
                    Cancelled: ${item.cancellation_date}
                </div>
                <div class="text-danger small fw-bold">
                    Refund: Rs. ${Number(item.refund_amount).toLocaleString()}
                </div>
            </div>`;
        });
        html += `</div>`;
    }

    // ── YELLOW: Due within 7 days ──────────────────────
    if (data.yellow && data.yellow.due_soon.length > 0) {
        html += `<div class="alert-section mt-2">
            <div class="alert-section-title text-warning">
                <i class="bi bi-clock-fill"></i>
                Due Within 7 Days (${data.yellow.due_soon.length})
            </div>`;
        data.yellow.due_soon.forEach(item => {
            html += `<div class="alert-item border-start border-warning border-3 ps-2 mb-2">
                <div class="fw-semibold">${item.customer_name}</div>
                <div class="text-muted small">
                    Plot ${item.plot_no} — ${item.block_name} |
                    Installment #${item.installment_no} |
                    Due: ${item.due_date}
                </div>
                <div class="text-warning small fw-bold">
                    Rs. ${Number(item.amount).toLocaleString()}
                </div>
            </div>`;
        });
        html += `</div>`;
    }

    // ── YELLOW: Landlord payments due ─────────────────
    if (data.yellow && data.yellow.landlord_due.length > 0) {
        html += `<div class="alert-section mt-2">
            <div class="alert-section-title text-warning">
                <i class="bi bi-building"></i>
                Landlord Payments Due (${data.yellow.landlord_due.length})
            </div>`;
        data.yellow.landlord_due.forEach(item => {
            html += `<div class="alert-item border-start border-warning border-3 ps-2 mb-2">
                <div class="fw-semibold">${item.landlord_name}</div>
                <div class="text-muted small">
                    ${item.contract_title} |
                    Installment #${item.installment_no} |
                    Due: ${item.due_date}
                </div>
                <div class="text-warning small fw-bold">
                    Rs. ${Number(item.amount).toLocaleString()}
                </div>
            </div>`;
        });
        html += `</div>`;
    }

    // ── YELLOW: Pending cheques ────────────────────────
    if (data.yellow && data.yellow.pending_cheques.length > 0) {
        html += `<div class="alert-section mt-2">
            <div class="alert-section-title text-warning">
                <i class="bi bi-journal-check"></i>
                Cheques Pending Clearance (${data.yellow.pending_cheques.length})
            </div>`;
        data.yellow.pending_cheques.forEach(item => {
            html += `<div class="alert-item border-start border-warning border-3 ps-2 mb-2">
                <div class="fw-semibold">${item.customer_name}</div>
                <div class="text-muted small">
                    Cheque #${item.cheque_no} — ${item.cheque_bank} |
                    Date: ${item.cheque_date}
                </div>
                <div class="text-warning small fw-bold">
                    Rs. ${Number(item.amount).toLocaleString()}
                </div>
            </div>`;
        });
        html += `</div>`;
    }

    // ── Empty state ────────────────────────────────────
    if (html === '') {
        html = `<div class="text-center text-muted py-4">
            <i class="bi bi-check-circle-fill text-success fs-3"></i>
            <div class="mt-2">No alerts. All clear!</div>
        </div>`;
    }

    panel.innerHTML = html;
}


// ======================================================
// INIT — run on page load then repeat every 60s
// ======================================================

document.addEventListener('DOMContentLoaded', function () {
    fetchAlerts();
    setInterval(fetchAlerts, REFRESH_MS);
});

File: REDS/flask_api/routes/alerts.py
# ======================================================
# FILE     : flask_api/routes/alerts.py
# PURPOSE  : Returns alert/notification data as JSON
#            Called by dashboard JavaScript
# ======================================================

from flask import Blueprint, jsonify
from datetime import date, timedelta
import sqlite3
import os

alerts_bp = Blueprint('alerts', __name__)

# ── Path to Django's SQLite database ──────────────────
BASE_DIR   = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
DB_PATH    = os.path.join(BASE_DIR, 'db.sqlite3')


def get_db():
    """Open a read-only connection to the shared SQLite DB."""
    conn = sqlite3.connect(f'file:{DB_PATH}?mode=ro', uri=True)
    conn.row_factory = sqlite3.Row
    return conn


# ======================================================
# MAIN ALERTS ENDPOINT
# GET /api/alerts/
# Returns counts and detail lists for dashboard
# ======================================================

@alerts_bp.route('/api/alerts/')
def get_alerts():
    today      = date.today().isoformat()
    soon       = (date.today() + timedelta(days=7)).isoformat()
    soon_30    = (date.today() + timedelta(days=30)).isoformat()

    conn = get_db()
    cur  = conn.cursor()

    # ── RED: Overdue customer installments ────────────
    cur.execute("""
        SELECT
            pp.id,
            pp.installment_no,
            pp.due_date,
            pp.amount,
            c.name  AS customer_name,
            pl.plot_no,
            bl.name AS block_name,
            t.name  AS town_name
        FROM sales_paymentplan pp
        JOIN sales_booking b   ON b.id = pp.booking_id
        JOIN customers_customer c ON c.id = b.customer_id
        JOIN development_plot pl  ON pl.id = b.plot_id
        JOIN development_block bl ON bl.id = pl.block_id
        JOIN development_town  t  ON t.id  = bl.town_id
        WHERE pp.status IN ('PENDING','PARTIAL')
          AND pp.due_date < ?
          AND b.status = 'ACTIVE'
          AND b.is_deleted = 0
        ORDER BY pp.due_date ASC
        LIMIT 20
    """, (today,))
    overdue_installments = [dict(r) for r in cur.fetchall()]

    # ── RED: Bounced cheques ───────────────────────────
    cur.execute("""
        SELECT
            r.id,
            r.receipt_no,
            r.cheque_no,
            r.cheque_bank,
            r.amount,
            c.name AS customer_name
        FROM sales_receipt r
        JOIN sales_booking b      ON b.id = r.booking_id
        JOIN customers_customer c ON c.id = b.customer_id
        WHERE r.cheque_status = 'BOUNCED'
          AND r.is_deleted = 0
        ORDER BY r.created_at DESC
        LIMIT 20
    """)
    bounced_cheques = [dict(r) for r in cur.fetchall()]

    # ── RED: Pending refunds (cancelled bookings) ──────
    cur.execute("""
        SELECT
            cn.id,
            cn.refund_amount,
            cn.cancellation_date,
            c.name AS customer_name,
            pl.plot_no,
            bl.name AS block_name
        FROM sales_cancellation cn
        JOIN sales_booking b      ON b.id = cn.booking_id
        JOIN customers_customer c ON c.id = b.customer_id
        JOIN development_plot pl  ON pl.id = b.plot_id
        JOIN development_block bl ON bl.id = pl.block_id
        WHERE cn.refund_amount > 0
        ORDER BY cn.cancellation_date DESC
        LIMIT 20
    """)
    pending_refunds = [dict(r) for r in cur.fetchall()]

    # ── YELLOW: Installments due within 7 days ─────────
    cur.execute("""
        SELECT
            pp.id,
            pp.installment_no,
            pp.due_date,
            pp.amount,
            c.name  AS customer_name,
            pl.plot_no,
            bl.name AS block_name
        FROM sales_paymentplan pp
        JOIN sales_booking b      ON b.id = pp.booking_id
        JOIN customers_customer c ON c.id = b.customer_id
        JOIN development_plot pl  ON pl.id = b.plot_id
        JOIN development_block bl ON bl.id = pl.block_id
        WHERE pp.status IN ('PENDING','PARTIAL')
          AND pp.due_date >= ?
          AND pp.due_date <= ?
          AND b.status = 'ACTIVE'
          AND b.is_deleted = 0
        ORDER BY pp.due_date ASC
        LIMIT 20
    """, (today, soon))
    due_soon = [dict(r) for r in cur.fetchall()]

    # ── YELLOW: Landlord payments due within 30 days ───
    cur.execute("""
        SELECT
            cps.id,
            cps.installment_no,
            cps.due_date,
            cps.amount,
            ll.name AS landlord_name,
            lc.title AS contract_title
        FROM land_contractpaymentschedule cps
        JOIN land_landcontract lc ON lc.id = cps.contract_id
        JOIN land_landlord     ll ON ll.id = lc.landlord_id
        WHERE cps.status IN ('PENDING','PARTIAL')
          AND cps.due_date >= ?
          AND cps.due_date <= ?
        ORDER BY cps.due_date ASC
        LIMIT 20
    """, (today, soon_30))
    landlord_due = [dict(r) for r in cur.fetchall()]

    # ── YELLOW: Cheques pending clearance > 7 days ─────
    cur.execute("""
        SELECT
            r.id,
            r.receipt_no,
            r.cheque_no,
            r.cheque_bank,
            r.cheque_date,
            r.amount,
            c.name AS customer_name
        FROM sales_receipt r
        JOIN sales_booking b      ON b.id = r.booking_id
        JOIN customers_customer c ON c.id = b.customer_id
        WHERE r.payment_mode = 'CHEQUE'
          AND (r.cheque_status = 'PENDING' OR r.cheque_status = '')
          AND r.cheque_date <= ?
          AND r.is_deleted = 0
        ORDER BY r.cheque_date ASC
        LIMIT 20
    """, ((date.today() - timedelta(days=7)).isoformat(),))
    pending_cheques = [dict(r) for r in cur.fetchall()]

    # ── GREEN: Plots available after cancellation ──────
    cur.execute("""
        SELECT COUNT(*) AS cnt
        FROM development_plot
        WHERE status = 'AVAILABLE'
          AND is_deleted = 0
    """)
    available_plots = cur.fetchone()['cnt']

    # ── GREEN: Receipts entered today ──────────────────
    cur.execute("""
        SELECT COUNT(*) AS cnt
        FROM sales_receipt
        WHERE DATE(created_at) = ?
          AND is_deleted = 0
    """, (today,))
    receipts_today = cur.fetchone()['cnt']

    conn.close()

    # ── Build summary counts ───────────────────────────
    red_count    = len(overdue_installments) + len(bounced_cheques) + len(pending_refunds)
    yellow_count = len(due_soon) + len(landlord_due) + len(pending_cheques)

    return jsonify({
        'red_count':    red_count,
        'yellow_count': yellow_count,

        'red': {
            'overdue_installments': overdue_installments,
            'bounced_cheques':      bounced_cheques,
            'pending_refunds':      pending_refunds,
        },

        'yellow': {
            'due_soon':       due_soon,
            'landlord_due':   landlord_due,
            'pending_cheques': pending_cheques,
        },

        'green': {
            'available_plots': available_plots,
            'receipts_today':  receipts_today,
        },
    })